corgidrp/detector.py

```diff
@@ -520,7 +528,149 @@ def calc_sat_fwc(emgain_arr,fwcpp_arr,fwcem_arr,sat_thresh):
     Returns:
         np.array: lowest full well capacity saturation threshold for frames
     """
-    possible_sat_fwcs_arr = np.append((emgain_arr * fwcpp_arr)[:,np.newaxis], fwcem_arr[:,np.newaxis],axis=1)
-    sat_fwcs = sat_thresh * np.min(possible_sat_fwcs_arr,axis=1)
-
-    return sat_fwcs
+	possible_sat_fwcs_arr = np.append((emgain_arr * fwcpp_arr)[:,np.newaxis], fwcem_arr[:,np.newaxis],axis=1)
+	sat_fwcs = sat_thresh * np.min(possible_sat_fwcs_arr,axis=1)
+
+	return sat_fwcs
+
+def flatfield_residuals(images,planims=None):
+    """Turn this dataset of image frames of neptune or uranus and create matched filters and estimate residuals after 
+     dividing from matched filters
+
+     Args:
+    	images (np.array): 2D array of cropped neptune or uranus image frames
+    	planims (str): name of the planet uranus or neptune
+
+     Returns:
+    	matched_residuals (np.array): residual image frames of neptune or uranus divided by matched filter
+	"""
+    if planims=='neptune' or planims=='uranus' :
+        raster_images = np.array([image[2] for image in images])
+        masks = np.array([image[3] for image in images]) 
+        images_split = np.array(np.split(np.array(raster_images),3))
+        matched_filters = np.array([np.nanmedian(np.stack(images_split[i],2),2) for i in np.arange(0,len(images_split))])
+        matched_filters_smooth = [gauss(matched_filters[i],3) for i in range(len(matched_filters))] 
+        matched_residuals=[];
+        for j in range(len(raster_images)):
+            matched_residuals.append(raster_images[j]/matched_filters_smooth[int(np.floor(j//(len(raster_images)//len(matched_filters_smooth))))])
+    return(matched_residuals)
+	    
+def combine_rasters(residual_images,cent=None,planet=None,band=None):
+    """combine the dataset of residual image frames of neptune or uranus and create flat field 
+    	and associated error
+
+    	Args:
+        	residual_images (np.array): residual image frames of neptune and uranus 
+        	cent (np.array): centroid of the image frames
+        	planet (str):   name of the planet neptune or uranus
+        	band (str):  band of the observation band1 or band4
+        	
+    	Returns:
+        	full_qe (np.array): ideal flat field
+        	full_residuals (np.array): flat field from uranus or neptune images
+        	percens_resel (np.array): percentage error of the flat field
+    """
+    
+    n = 420
```

Comment:
It looks like we have some hard coded variables here: `n`, `rad`, `rad1, and `rad4`. Do we think some of them may need to be tuned or tweaked with the real data? If so, we should expose them as variables we can change via keywords. If not, it's ok to leave them be. I just want to make sure we think about it.

Reply:
I have made these the keywords for the create_onsky_flatfield function.